File: src/train/rolloutBuffer.ts
import * as tf from '@tensorflow/tfjs'
import type { ActionInput, ActionMask, Observation, StepResult } from '../env'
import { Env, buildActionMasksForEnv } from '../env'
import { deriveRng, isGameOver, randInt } from '../sim'
import type { ActorCriticModel } from './network'
import { sampleMaskedCategorical } from './actionSampling'
import { computeGAE } from './gae'
import type { RolloutBatch, RolloutStep } from './types'

/**
 * 複数の`Env`インスタンスをロックステップで駆動する永続状態(バッチ化・単一プロセス内 —
 * 実スレッドではない。`Env`自身が「並べて多数生成できる」ことを前提に設計されている §1)。
 * `collectRollout`を反復呼び出しする間、各envの「現在のエピソード」を跨いで保持する
 * (呼び出しごとに全envをリセットすると、直前のロールアウトの続きからではなく
 * 常にエピソード冒頭からしかデータを取れなくなってしまう)。
 */
export interface RolloutState {
  envs: Env[]
  /** env毎の、リセットのたびに1ずつ増えるカウンタ。ユニットIDはリセットのたびに再利用される
   * ため(entities.tsが毎回0から振り直す)、セグメントを`(envIndex, episodeId, unitId)`で
   * 一意化するのに使う。 */
  episodeIds: number[]
  /** env毎の、現在アクティブなエージェントに対する直近の観測。 */
  obs: Record<number, Observation>[]
  /** env毎の、現在アクティブなエージェントに対する直近の行動マスク。 */
  masks: Record<number, ActionMask>[]
}

function deriveEnvSeed(baseSeed: number, envIndex: number, episodeId: number): number {
  const rng = deriveRng(baseSeed, `rollout:env${envIndex}:ep${episodeId}`)
  return randInt(rng, 2 ** 31)
}

export function createRolloutState(envs: Env[], baseSeed: number): RolloutState {
  const episodeIds = envs.map(() => 0)
  const obs: Record<number, Observation>[] = []
  const masks: Record<number, ActionMask>[] = []
  envs.forEach((env, i) => {
    obs.push(env.reset(deriveEnvSeed(baseSeed, i, 0)))
    masks.push(buildActionMasksForEnv(env))
  })
  return { envs, episodeIds, obs, masks }
}

interface RawRecord {
  envIndex: number
  episodeId: number
  unitId: number
  obs: number[]
  moveMask: boolean[]
  attackMask: boolean[]
  moveAction: number
  attackAction: number
  oldLogProb: number
  value: number
  reward: number
  /** Envの`terminations[unitId]`(本当に死んだか)。打ち切り/ロールアウト末端での継続中はfalse。 */
  terminated: boolean
}

function segmentKey(envIndex: number, episodeId: number, unitId: number): string {
  return `${envIndex}:${episodeId}:${unitId}`
}

function computeBootstrapValues(
  model: ActorCriticModel,
  obsByUnit: Record<number, Observation>,
  unitIds: number[],
): number[] {
  if (unitIds.length === 0) return []
  return tf.tidy(() => {
    const obsT = tf.tensor2d(unitIds.map((id) => obsByUnit[id].vector))
    const { value } = model.forward(obsT)
    return Array.from(value.dataSync())
  })
}

export interface CollectRolloutOptions {
  rolloutLength: number
  gamma: number
  lambda: number
  /** env自動リセットのseed導出に使う基点。呼び出しごとに変える必要はない —
   * `episodeIds`が呼び出しを跨いで単調増加し続けるため、同じ`baseSeed`でも
   * リセットのたびに異なるseedが導出される。 */
  baseSeed: number
}

/**
 * `rolloutLength`意思決定ステップ分、全envをロックステップで進めてデータを収集する。
 * 各意思決定ステップで、全envの現在アクティブな`(env,unit)`ペアを1つのバッチにまとめて
 * 1回の`model.forward`で処理する(§1の「複数環境インスタンスを同一プロセスで並列生成」を
 * バッチ化フォワードパスとして実現)。envが空になる/ゲーム終了したら即座にリセットし、
 * 収集が滞らないようにする。
 */
export function collectRollout(
  state: RolloutState,
  model: ActorCriticModel,
  opts: CollectRolloutOptions,
): RolloutBatch {
  const { envs, episodeIds, obs, masks } = state
  const numEnvs = envs.length

  const records: RawRecord[] = []
  const bootstrapValues = new Map<string, number>()
  const episodeReturnAccum: Map<number, number>[] = Array.from({ length: numEnvs }, () => new Map())
  const episodeReturns: number[] = []

  for (let step = 0; step < opts.rolloutLength; step++) {
    const active: { envIndex: number; unitId: number }[] = []
    const obsRows: number[][] = []
    const moveMaskRows: number[][] = []
    const attackMaskRows: number[][] = []

    for (let e = 0; e < numEnvs; e++) {
      for (const unitId of envs[e].agents) {
        const o = obs[e][unitId]
        const m = masks[e][unitId]
        if (!o || !m) continue
        active.push({ envIndex: e, unitId })
        obsRows.push(o.vector)
        moveMaskRows.push(m.move.map((b) => (b ? 1 : 0)))
        attackMaskRows.push(m.attack.map((b) => (b ? 1 : 0)))
      }
    }
    if (active.length === 0) break // shouldn't happen given per-step auto-reset below; defensive only

    const { moveActions, attackActions, oldLogProbs, values } = tf.tidy(() => {
      const obsT = tf.tensor2d(obsRows)
      const moveMaskT = tf.tensor2d(moveMaskRows)
      const attackMaskT = tf.tensor2d(attackMaskRows)
      const { moveLogits, attackLogits, value } = model.forward(obsT)
      const moveSample = sampleMaskedCategorical(moveLogits, moveMaskT)
      const attackSample = sampleMaskedCategorical(attackLogits, attackMaskT)
      const jointLogProb = tf.add(moveSample.logProbs, attackSample.logProbs)
      return {
        moveActions: Array.from(moveSample.actions.dataSync()),
        attackActions: Array.from(attackSample.actions.dataSync()),
        oldLogProbs: Array.from(jointLogProb.dataSync()),
        values: Array.from(value.reshape([value.shape[0]]).dataSync()),
      }
    })

    const perEnvActions: Record<number, ActionInput>[] = Array.from({ length: numEnvs }, () => ({}))
    active.forEach((a, i) => {
      perEnvActions[a.envIndex][a.unitId] = { move: moveActions[i], attack: attackActions[i] }
    })

    const stepResults: StepResult[] = envs.map((env, e) => env.step(perEnvActions[e]))

    active.forEach((a, i) => {
      const sr = stepResults[a.envIndex]
      const reward = sr.rewards[a.unitId] ?? 0
      const terminated = sr.terminations[a.unitId] ?? false
      records.push({
        envIndex: a.envIndex,
        episodeId: episodeIds[a.envIndex],
        unitId: a.unitId,
        obs: obsRows[i],
        moveMask: moveMaskRows[i].map((v) => v === 1),
        attackMask: attackMaskRows[i].map((v) => v === 1),
        moveAction: moveActions[i],
        attackAction: attackActions[i],
        oldLogProb: oldLogProbs[i],
        value: values[i],
        reward,
        terminated,
      })
      const accum = episodeReturnAccum[a.envIndex]
      accum.set(a.unitId, (accum.get(a.unitId) ?? 0) + reward)
    })

    for (let e = 0; e < numEnvs; e++) {
      const sr = stepResults[e]
      const postStepObs = sr.observations
      const postStepMasks: Record<number, ActionMask> = {}
      for (const [unitIdStr, info] of Object.entries(sr.infos)) {
        postStepMasks[Number(unitIdStr)] = info.actionMask
      }

      const stillAliveAgents = envs[e].agents // reflects state after env.step() above
      // `maxTicks`truncation only sets per-unit `truncations` flags (Env itself never auto-stops);
      // without also treating that as a reset trigger here, an `EnvOptions.maxTicks` cap would be
      // silently ignored and episodes could run indefinitely past it.
      const truncatedThisStep = Object.values(sr.truncations).some(Boolean)
      const needsReset = stillAliveAgents.length === 0 || isGameOver(envs[e].state) || truncatedThisStep

      if (needsReset && stillAliveAgents.length > 0) {
        // Natural episode end (win/elimination/truncation) while these units are still alive —
        // not a real death, so they need a real bootstrap value, computed now before their
        // observations become unreachable (the env is about to reset to a brand-new episode).
        const bootVals = computeBootstrapValues(model, postStepObs, stillAliveAgents)
        stillAliveAgents.forEach((unitId, idx) => {
          bootstrapValues.set(segmentKey(e, episodeIds[e], unitId), bootVals[idx])
        })
      }

      if (needsReset) {
        episodeReturns.push(...episodeReturnAccum[e].values())
        episodeReturnAccum[e] = new Map()
        episodeIds[e]++
        obs[e] = envs[e].reset(deriveEnvSeed(opts.baseSeed, e, episodeIds[e]))
        masks[e] = buildActionMasksForEnv(envs[e])
      } else {
        obs[e] = postStepObs
        masks[e] = postStepMasks
      }
    }
  }

  // Rollout window ended while some units are still mid-life (never terminated, never hit a
  // mid-rollout reset) — bootstrap them too, using the same map, so every open segment resolves.
  for (let e = 0; e < numEnvs; e++) {
    const aliveNow = envs[e].agents
    if (aliveNow.length === 0) continue
    const bootVals = computeBootstrapValues(model, obs[e], aliveNow)
    aliveNow.forEach((unitId, idx) => {
      bootstrapValues.set(segmentKey(e, episodeIds[e], unitId), bootVals[idx])
    })
  }

  const grouped = new Map<string, RawRecord[]>()
  for (const r of records) {
    const key = segmentKey(r.envIndex, r.episodeId, r.unitId)
    const arr = grouped.get(key)
    if (arr) arr.push(r)
    else grouped.set(key, [r])
  }

  const steps: RolloutStep[] = []
  for (const [key, recs] of grouped) {
    const terminal = recs[recs.length - 1].terminated
    const bootstrapValue = terminal ? 0 : (bootstrapValues.get(key) ?? 0)
    const { advantages, returns } = computeGAE(
      { rewards: recs.map((r) => r.reward), values: recs.map((r) => r.value), bootstrapValue, terminal },
      opts.gamma,
      opts.lambda,
    )
    recs.forEach((r, i) => {
      steps.push({
        obs: r.obs,
        moveMask: r.moveMask,
        attackMask: r.attackMask,
        moveAction: r.moveAction,
        attackAction: r.attackAction,
        oldLogProb: r.oldLogProb,
        value: r.value,
        advantage: advantages[i],
        return: returns[i],
      })
    })
  }

  return { steps, episodeReturns }
}

File: training-py/mayhem_rl/rollout_buffer.py
"""Vectorized rollout collection driving a `WorkerPool` (the bridge's equivalent of
src/train/rolloutBuffer.ts, which drives in-process `Env` instances directly). Mirrors that
file's batching-by-flattening-alive-(env,unit)-pairs approach, its auto-reset handling, and its
segment-per-life GAE bookkeeping — adapted so that the actual env-stepping happens over the wire
via `WorkerPool`, and every forward pass (including bootstrap-value-only passes) happens here in
Python, never inside a Node worker (the bridge's non-negotiable constraint).
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Any

import torch

from .action_sampling import sample_masked_categorical
from .bridge.worker_pool import WorkerPool
from .gae import GaeSegment, compute_gae
from .network import ActorCriticNetwork
from .ppo import RolloutStep


def _initial_seed(base_seed: int, global_index: int) -> int:
    """Only used for the *explicit* initial reset — auto-resets that happen organically during
    `step()` are seeded by the worker itself (via its own `baseSeed` + `deriveRng`, see
    envWorker.ts), so this doesn't need to match that convention; it just needs to be
    deterministic and distinct per env for reproducibility within a Python run."""
    rng = random.Random(f"{base_seed}:{global_index}:reset0")
    return rng.randrange(2**31)


@dataclass
class RolloutState:
    pool: WorkerPool
    base_seed: int
    episode_ids: list[int]
    # per global env index: {unitId: {"observation": [...], "visibleEnemyIds": [...], "actionMask": {...}}}
    agents_by_env: list[dict[int, dict[str, Any]]]


def create_rollout_state(pool: WorkerPool, base_seed: int) -> RolloutState:
    seeds = {i: _initial_seed(base_seed, i) for i in range(pool.num_envs)}
    reset_result = pool.reset(seeds)

    episode_ids = [0] * pool.num_envs
    agents_by_env: list[dict[int, dict[str, Any]]] = [dict() for _ in range(pool.num_envs)]
    for global_index, env_result in reset_result.items():
        episode_ids[global_index] = env_result["episodeId"]
        agents_by_env[global_index] = {a["unitId"]: a for a in env_result["agents"]}

    return RolloutState(pool=pool, base_seed=base_seed, episode_ids=episode_ids, agents_by_env=agents_by_env)


@dataclass
class _RawRecord:
    env_index: int
    episode_id: int
    unit_id: int
    obs: list[float]
    move_mask: list[bool]
    attack_mask: list[bool]
    ability_mask: list[bool]
    move_action: int
    attack_action: int
    ability_action: int
    old_log_prob: float
    value: float
    reward: float
    terminated: bool


@dataclass
class RolloutBatch:
    steps: list[RolloutStep]
    episode_returns: list[float]


def _compute_values(network: ActorCriticNetwork, device: torch.device, obs_list: list[list[float]]) -> list[float]:
    if not obs_list:
        return []
    with torch.no_grad():
        obs_t = torch.tensor(obs_list, dtype=torch.float32, device=device)
        _, _, _, value = network(obs_t)
    return value.squeeze(-1).cpu().tolist()


def collect_rollout(
    state: RolloutState,
    network: ActorCriticNetwork,
    device: torch.device,
    rollout_length: int,
    gamma: float,
    lam: float,
    action_rng: random.Random,
) -> RolloutBatch:
    pool = state.pool
    records: list[_RawRecord] = []
    bootstrap_values: dict[tuple[int, int, int], float] = {}
    episode_return_accum: list[dict[int, float]] = [dict() for _ in range(pool.num_envs)]
    episode_returns: list[float] = []

    for _step in range(rollout_length):
        active: list[tuple[int, int]] = []  # (env_index, unit_id)
        obs_rows: list[list[float]] = []
        move_mask_rows: list[list[bool]] = []
        attack_mask_rows: list[list[bool]] = []
        ability_mask_rows: list[list[bool]] = []

        for env_index in range(pool.num_envs):
            for unit_id, agent in state.agents_by_env[env_index].items():
                active.append((env_index, unit_id))
                obs_rows.append(agent["observation"])
                move_mask_rows.append(agent["actionMask"]["move"])
                attack_mask_rows.append(agent["actionMask"]["attack"])
                ability_mask_rows.append(agent["actionMask"]["ability"])

        if not active:
            break  # defensive; shouldn't happen given per-step auto-reset, mirrors rolloutBuffer.ts

        obs_t = torch.tensor(obs_rows, dtype=torch.float32, device=device)
        move_mask_t = torch.tensor(
            [[1.0 if b else 0.0 for b in row] for row in move_mask_rows], dtype=torch.float32, device=device
        )
        attack_mask_t = torch.tensor(
            [[1.0 if b else 0.0 for b in row] for row in attack_mask_rows], dtype=torch.float32, device=device
        )
        ability_mask_t = torch.tensor(
            [[1.0 if b else 0.0 for b in row] for row in ability_mask_rows], dtype=torch.float32, device=device
        )

        seed = action_rng.randrange(2**31)
        gen_move = torch.Generator().manual_seed(seed)
        gen_attack = torch.Generator().manual_seed(seed + 1)
        gen_ability = torch.Generator().manual_seed(seed + 2)

        with torch.no_grad():
            move_logits, attack_logits, ability_logits, value_t = network(obs_t)
            # Action sampling always happens on CPU regardless of `device`: it's a handful of
            # floats per row (not the network forward pass itself), and `torch.multinomial`'s
            # `generator` argument must live on the same device as the tensor it samples from —
            # sampling on CPU sidesteps needing a CUDA generator at all.
            move_sample = sample_masked_categorical(move_logits.cpu(), move_mask_t.cpu(), generator=gen_move)
            attack_sample = sample_masked_categorical(attack_logits.cpu(), attack_mask_t.cpu(), generator=gen_attack)
            ability_sample = sample_masked_categorical(
                ability_logits.cpu(), ability_mask_t.cpu(), generator=gen_ability
            )
            joint_log_prob = move_sample.log_probs + attack_sample.log_probs + ability_sample.log_probs

        move_actions = move_sample.actions.tolist()
        attack_actions = attack_sample.actions.tolist()
        ability_actions = ability_sample.actions.tolist()
        old_log_probs = joint_log_prob.tolist()
        values = value_t.squeeze(-1).cpu().tolist()

        actions_by_env: dict[int, list[dict[str, Any]]] = {}
        for i, (env_index, unit_id) in enumerate(active):
            actions_by_env.setdefault(env_index, []).append(
                {
                    "unitId": unit_id,
                    "move": move_actions[i],
                    "attack": attack_actions[i],
                    "ability": ability_actions[i],
                }
            )

        step_result = pool.step(actions_by_env)

        for i, (env_index, unit_id) in enumerate(active):
            env_result = step_result[env_index]
            unit_result = next(u for u in env_result["units"] if u["unitId"] == unit_id)
            records.append(
                _RawRecord(
                    env_index=env_index,
                    episode_id=env_result["episodeId"],
                    unit_id=unit_id,
                    obs=obs_rows[i],
                    move_mask=move_mask_rows[i],
                    attack_mask=attack_mask_rows[i],
                    ability_mask=ability_mask_rows[i],
                    move_action=move_actions[i],
                    attack_action=attack_actions[i],
                    ability_action=ability_actions[i],
                    old_log_prob=old_log_probs[i],
                    value=values[i],
                    reward=unit_result["reward"],
                    terminated=unit_result["terminated"],
                )
            )
            accum = episode_return_accum[env_index]
            accum[unit_id] = accum.get(unit_id, 0.0) + unit_result["reward"]

        for env_index in range(pool.num_envs):
            env_result = step_result.get(env_index)
            if env_result is None:
                continue
            reset_field = env_result["reset"]
            if reset_field is None:
                state.agents_by_env[env_index] = {a["unitId"]: a for a in env_result["continuing"]}
                continue

            # Natural episode end (win/elimination/truncation) while these units were still
            # alive -- not a real death, so they need a real bootstrap value. Node already handed
            # us their final observation in `reset.bootstrap`; the forward pass happens here.
            bootstrap_obs = [b["observation"] for b in reset_field["bootstrap"]]
            bootstrap_vals = _compute_values(network, device, bootstrap_obs)
            for b, v in zip(reset_field["bootstrap"], bootstrap_vals):
                bootstrap_values[(env_index, env_result["episodeId"], b["unitId"])] = v

            episode_returns.extend(episode_return_accum[env_index].values())
            episode_return_accum[env_index] = {}

            state.episode_ids[env_index] = reset_field["newEpisodeId"]
            state.agents_by_env[env_index] = {a["unitId"]: a for a in reset_field["agents"]}

    # Rollout window ended while some units are still mid-life (never terminated, never hit a
    # mid-rollout reset) -- bootstrap them too, so every open segment resolves.
    for env_index in range(pool.num_envs):
        agents = state.agents_by_env[env_index]
        if not agents:
            continue
        unit_ids = list(agents.keys())
        values = _compute_values(network, device, [agents[u]["observation"] for u in unit_ids])
        for unit_id, v in zip(unit_ids, values):
            bootstrap_values[(env_index, state.episode_ids[env_index], unit_id)] = v

    grouped: dict[tuple[int, int, int], list[_RawRecord]] = {}
    for r in records:
        grouped.setdefault((r.env_index, r.episode_id, r.unit_id), []).append(r)

    steps: list[RolloutStep] = []
    for key, recs in grouped.items():
        terminal = recs[-1].terminated
        bootstrap_value = 0.0 if terminal else bootstrap_values.get(key, 0.0)
        result = compute_gae(
            GaeSegment(
                rewards=[r.reward for r in recs],
                values=[r.value for r in recs],
                bootstrap_value=bootstrap_value,
                terminal=terminal,
            ),
            gamma,
            lam,
        )
        for i, r in enumerate(recs):
            steps.append(
                RolloutStep(
                    obs=r.obs,
                    move_mask=r.move_mask,
                    attack_mask=r.attack_mask,
                    ability_mask=r.ability_mask,
                    move_action=r.move_action,
                    attack_action=r.attack_action,
                    ability_action=r.ability_action,
                    old_log_prob=r.old_log_prob,
                    value=r.value,
                    advantage=result.advantages[i],
                    return_=result.returns[i],
                )
            )

    return RolloutBatch(steps=steps, episode_returns=episode_returns)
